carto/sql.py

```diff
@@ -226,3 +243,200 @@ def cancel(self, job_id):
+
+    def _read_in_chunks(self, file_object, chunk_size=DEFAULT_CHUNK_SIZE):
+        while True:
+            data = file_object.read(chunk_size)
+            if not data:
+                break
+            yield data
+
+    def _compress_chunks(self, chunk_generator, compression_level):
+        zlib_mode = 16 + zlib.MAX_WBITS
+        compressor = zlib.compressobj(compression_level, zlib.DEFLATED, zlib_mode)
+        for chunk in chunk_generator:
+            compressed_chunk = compressor.compress(chunk)
+            if len(compressed_chunk) > 0:
+                yield compressed_chunk
+        yield compressor.flush()
+
+
+    def copyfrom(self, query, iterable_data, compress=True, compression_level=DEFAULT_COMPRESSION_LEVEL):
+        """
+        Gets data from an iterable object into a table
+
+        :param query: The "COPY table_name [(column_name[, ...])]
+                           FROM STDIN [WITH(option[,...])]" query to execute
+        :type query: str
+
+        :param iterable_data: An object that can be iterated
+                              to retrieve the data
+        :type iterable_data: object
+
+        :return: Response data as json
+        :rtype: str
+
+        :raise CartoException:
+        """
+        url = self.api_url + '/copyfrom'
+        headers = {
+            'Content-Type': 'application/octet-stream',
+            'Transfer-Encoding': 'chunked'
+        }
+        params={'api_key': self.api_key, 'q': query}
+
+        if compress:
+            headers['Content-Encoding'] = 'gzip'
+            _iterable_data = self._compress_chunks(iterable_data, compression_level)
+        else:
+            _iterable_data = iterable_data
+
+        try:
+            response = self.client.send(url,
+                                        http_method='POST',
+                                        params=params,
+                                        data=_iterable_data,
+                                        headers=headers,
+                                        stream=True)
+            response_json = self.client.get_response_data(response)
+        except Exception as e:
+            raise CartoException(e)
+
+        return response_json
+
+    def copyfrom_file_object(self, query, file_object, compress=True, compression_level=DEFAULT_COMPRESSION_LEVEL):
+        """
+        Gets data from a readable file object into a table
+
+        :param query: The "COPY table_name [(column_name[, ...])]
+                           FROM STDIN [WITH(option[,...])]" query to execute
+        :type query: str
+
+        :param file_object: A file-like object.
+                            Normally the return value of open('file.ext', 'rb')
+        :type file_object: file
+
+        :return: Response data as json
+        :rtype: str
+
+        :raise CartoException:
+        """
+        chunk_generator = self._read_in_chunks(file_object)
+        return self.copyfrom(query, chunk_generator, compress)
```

Comment:
Silly mistake: there is no use adding the `compression_level` parameter if you don't pass it down the call stack :sweat_smile: 

Same in the next method.

Reply:
shame on me :slightly_frowning_face: 

I will fix it ASAP. Thanks for pointing it out :slightly_smiling_face: